### PR TITLE
Allow using CORS filtered image responses as WebGL textures

### LINKED ISSUE #24330: WebGL: Allow TexImage from a different origin if CORS used
https://github.com/servo/servo/blob/cf725fc3f59bc1b94e3b35ab4cfe60b3d41559cc/components/script/dom/webglrenderingcontext.rs#L567

### LINKED ISSUE #24368: Image cache needs to be keyed on more attributes
According to https://html.spec.whatwg.org/multipage/images.html#list-of-available-images we need to consider:
* the URL
* the source element's CORS attribute
* an origin (optionally)

### PR DESCRIPTION
**Pull request by jdm**
More specifically, this makes the "is this image same origin?" check consider the CORS status of the original response, rather than relying on an overly-strict "is this image's response's URL same-origin with a particular global?" check. To do this, we make the image cache double keyed based on the requested URL as well as the requesting origin, and store the CORS status of the eventual response with the final image that eventually gets sent to the HTMLImageElement consumer.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24330 and fix #24368
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/24340)

### CODE REVIEW

**Comment by jdm** [comment]
This change makes the floor appear in http://paulrouget.com/test/bbjs/.

**Comment by jdm** [comment]
@asajeffrey Your thoughts on the second commit here would be interesting.

**Comment by paulrouget** [comment]
> This change makes the floor appear in http://paulrouget.com/test/bbjs/

I changed the demo to use same-domain images.

I tested the PR and it doesn't fix it.

**Comment by paulrouget** [comment]
The demo has moved to http://paulrouget.com/test/bbjs/basic

**Comment by jdm** [comment]
I tested the change and the floor appeared for me :<

**Comment by paulrouget** [comment]
> I tested the change and the floor appeared for me :<

It's possible that I changed the demo as you were writing the PR. Or maybe I haven't properly tested your patch.

File: components/net_traits/image_cache.rs [hunk 1]
```diff
@@ -121,3 +122,9 @@ pub trait ImageCache: Sync + Send {
     /// Inform the image cache about a response for a pending request.
     fn notify_pending_response(&self, id: PendingImageId, action: FetchResponseMsg);
 }
+
+#[derive(Clone, Copy, Debug, Deserialize, MallocSizeOf, PartialEq, Serialize)]
+pub enum CorsStatus {
+    Safe,
```
Comment (Manishearth) [inline components/net_traits/image_cache.rs]: Documentation on what exactly each variant means; what cases they cover.

**Review by Manishearth — APPROVED** [review]

File: components/net/image_cache.rs [hunk 2]
```diff
@@ -472,15 +500,26 @@ impl ImageCache for ImageCacheImpl {
...
-                    Ok(meta) => Some(match meta {
-                        FetchMetadata::Unfiltered(m) => m,
-                        FetchMetadata::Filtered { unsafe_, .. } => unsafe_,
-                    }),
-                    Err(_) => None,
+                let (cors_status, metadata) = match response {
+                    Ok(meta) => match meta {
+                        FetchMetadata::Unfiltered(m) => (CorsStatus::Safe, Some(m)),
+                        FetchMetadata::Filtered { unsafe_, filtered } => (
+                            match filtered {
```
[the remark this answers is not in the document]
Comment (jdm) [reply components/net/image_cache.rs]: Can you give an example of what you mean?

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/net/image_cache.rs]: The `Some(unsafe_)` line specifically looks like a new match arm

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/net/image_cache.rs]: I'm basically suggesting

```rust
FetchMetadata::Filtered { unsafe_, filtered } => {
    (
        match filtered {...},
        Some(...)
    )
}
```

**Comment by jdm** [comment]
@paulrouget I would expect these changes to work now. I verified that https://www.babylonjs-playground.com/frame.html#YDO1F#75 had a security error before the latest commit and now renders correctly.

File: components/net/image_cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdm) [reply components/net/image_cache.rs]: Rustfmt is unhappy with your suggestion and just reverts it.

**Comment by paulrouget** [comment]
It works.

**Comment by jdm** [comment]
Ok, this ready for review. There's even a new test that exposed #24368.

File: components/script/dom/htmlmediaelement.rs [hunk 3]
```diff
@@ -822,16 +822,9 @@ impl HTMLMediaElement {
             None => self.blob_url.borrow().as_ref().unwrap().clone(),
         };
 
-        let request = RequestBuilder::new(url.clone())
+        let cors_setting = cors_setting_for_element(self.upcast());
+        let request = create_a_potential_CORS_request(url.clone(), destination, cors_setting, None)
```
Comment (Manishearth) [inline components/script/dom/htmlmediaelement.rs]: yesss, the fewer `RequestBuilder::new`s we have the better

**Review by Manishearth — APPROVED** [review]

**Pull request merged**